File: py_tools/PET_Intensity.py
import os
import numpy as np
import SimpleITK as sitk
import csv

import pandas as pd
from tqdm import tqdm

def load_roi_info(roi_csv, header=False, delimiter=' '):
    """
    Load ROI information from a CSV file.
    Returns a dictionary whose key is the ROI value and whose value is (ROIid, ROIabbr, ROIname).
    """
    roi_info = {}
    with open(roi_csv, mode='r') as file:
        reader = csv.reader(file, delimiter=delimiter)
        if header:
            next(reader)  # Skip the header.
        for row in reader:
            roi_value = int(row[0])
            if len(row) <= 2:
                row.append(row[0])
            roi_info[roi_value] = (roi_value, row[1], row[2])
    return roi_info

def get_header(roi_info,Vox_mm3,SUVr):
    # Build the CSV header.
    header = ['Subject']
    for v in roi_info.keys():
        header.append(roi_info[v][1])
        if Vox_mm3:
            header.append(roi_info[v][1] + '_Vox_mm3')
        if SUVr:
            header.append(roi_info[v][1] + '_SUVr')
    return header

def get_pet_data(mask_path, nii_dir, output_csv, startswith='r', cerebellar=False, Vox_mm3=True, roi_info=None, subject_info=None, SUVr=False):
    print('Initializing...')
    ID = 'Subject ID'
    Injected_dose = 185  # MBq, FDG-PET reference in ADNI.
    cerebellar_ID = []  # 95~120
    s_info = None
    if SUVr:
        if subject_info:
            s_info = pd.read_csv(subject_info)
            s_info = s_info[s_info['Modality'] == 'PET']
            # Calculate SUVr normalized by the cerebellum. AAL3 cerebellar regions are:
            cerebellar_ID = np.arange(95, 121)  # 95~120
            # cerebellar_ID = np.arange(95, 113)  # 95~112
        else:
            print("Calculating the SUVr must input the subject_info(Weight) parameter")
            return



    with open(output_csv, mode='w', newline='') as file:
        writer = csv.writer(file)
        write_header = False
        # Get .nii or .nii.gz files.
        nii_files = [f for f in os.listdir(nii_dir) if
                     (f.endswith('.nii') or f.endswith('.nii.gz')) and f.startswith(startswith)]


        # Read the mask file and get unique values.
        mask = sitk.ReadImage(mask_path)
        mask_array = sitk.GetArrayFromImage(mask)
        unique_values = np.unique(mask_array)[np.unique(mask_array) != 0]  # Exclude 0.

        # Create a dictionary to store the indices for each value.
        indices_dict = {}
        for value in unique_values:
            indices = np.argwhere(mask_array == value)  # Get indices for each value.
            indices_dict[value] = indices  # Save the indices in the dictionary.

        # Check whether ROI length and mask length are consistent.
        if roi_info and len(unique_values) != len(roi_info) :
            print(f"\nROI_INFO length ({len(roi_info)}) and "
                  f"ROI.nii length ({len(unique_values)}) are inconsistent; unmatched ROI values will be 0")
        if roi_info is None:
            roi_info = {}
            for uv in unique_values:
                roi_info[uv] = (uv, 'ROI_'+str(uv), uv)
        header = get_header(roi_info,Vox_mm3,SUVr)
        writer.writerow(header)

        for nii_filename in tqdm(nii_files, desc="Processing NIfTI Files", unit="file"):
            ce_suv = 1
            weight = -1

            nii_path = os.path.join(nii_dir, nii_filename)
            nii = sitk.ReadImage(nii_path)
            nii_array = sitk.GetArrayFromImage(nii)

            sid, _ = os.path.splitext(nii_filename)
            sid = sid[len(startswith):]
            # sid = sid.removeprefix(startswith)

            if SUVr:
                weight_ = s_info[s_info[ID] == sid]['Weight']
                if len(weight_) >0:
                    weight = float(weight_.iloc[0]) * 1000
                else:
                    print(f'{sid} have not weight, SUVr May be 0')
                sum_ce = 0
                for ci in cerebellar_ID:
                    # Get all indices in the template that equal the current value.
                    indices = indices_dict[ci]
                    # ce_vox_num = len(indices)
                    ce_vox_in = nii_array[tuple(indices.T)]  # Transpose indices to match NumPy indexing.
                    # sum_ce += np.mean(ce_vox_in)
                    sum_ce += np.sum(ce_vox_in)
                if weight !=0:
                    ce_suv = sum_ce/(Injected_dose/weight)
                else:
                    weight = 80
                    ce_suv = sum_ce/(Injected_dose/weight)

            average_values = []
            Null_ROI = []
            for value in roi_info.keys():
                avg_intensity = 0
                vox_mm3 = 0
                roi_suvr = 0
                if (value not in unique_values):
                    Null_ROI.append(value)
                if (value not in unique_values) or (value in cerebellar_ID and not cerebellar):
                    pass
                else: # value in unique_values
                    # Get all indices in the template that equal the current value.
                    indices = indices_dict[value]
                    intensity = nii_array[tuple(indices.T)]  # Transpose indices to match NumPy indexing.
                    avg_intensity = np.mean(intensity)
                    if SUVr:
                        sum_intensity = np.sum(intensity)
                        roi_suv = sum_intensity/(Injected_dose/weight) if weight != -1 else 0
                        roi_suvr = roi_suv/ce_suv
                    vox_mm3 = len(indices)

                average_values.append(str(avg_intensity))
                if Vox_mm3:
                    average_values.append(str(vox_mm3))
                if SUVr:
                    average_values.append(str(roi_suvr))

            writer.writerow([sid] + average_values)
        print(f"Results written to CSV file: {output_csv}")
        print('Invalid ROI indices:',Null_ROI)


if __name__ == '__main__':
    # Usage example.
    ROI = r'tools\Reslice_aal3.nii' # ROI path.
    PET = rf'C:\Users\dongz\Desktop\adni_dataset\petjisuan'  # Directory for registered and normalized PET images.
    prefix = ''  # Corresponds to the registered PET filename prefix, for example wr for files starting with wr.
    # subject_Info is the CSV information file used when filtering data.
    subject_Info = rf'PET_processed.csv'
    roi_csv = 'tools/aal3.csv'  # ROI information file path.
    ROI_INFO = load_roi_info(roi_csv, header=True, delimiter=';')

    output_csv = 'PET_DATA.csv'  # Output CSV file path.
    # Besides intensity, configure whether to calculate cerebellar values, ROI volume, and SUVr.
    # SUVr requires subject_info containing patient weight.
    get_pet_data(ROI, PET, output_csv, startswith=prefix, cerebellar=True,
                 Vox_mm3=False, roi_info=ROI_INFO, subject_info=subject_Info, SUVr=True)

File: mat_tools/PET_Intensity.m
% NO Complete!!!
clc,clear;
roi = 'test/aal3.nii';
pet = 'test';
s_info = 'Info/Web_BOTH_MRI_PET.csv';
get_pet_data(roi,pet,'',s_info)
function get_pet_data(mask_path,pet_dir,roi_info,subject_info,Vox_mm3,SUVr,cerebellar,startswith,output_csv)
    if nargin < 4 subject_info = ''; end
    if nargin < 6 SUVr = 1; end
    if nargin < 5 Vox_mm3 = 1; end
    if nargin < 7 cerebellar = 0; end
    if nargin < 8 startswith = 'wr'; end
    if nargin < 9 output_csv = 'PET_DATA.csv'; end

    % Initialize parameters.
    ID = 'Subject ID';
    Injected_dose = 185;  % MBq, FDG-PET Reference ADNI
    cerebellar_ID = 95:112;  % Cerebellar region IDs.
    % cerebellar_ID = 95:120;  % Cerebellar region IDs.

    roi_info = load_roi_info('Template/ROITemplate/aal3.csv',1,';');
    header = get_header(roi_info, Vox_mm3, SUVr);
    if SUVr
        if ~isempty(subject_info)
            % s_info = readtable(subject_info,'VariableNamingRule','preserve');
            s_info = readtable(subject_info, 'VariableNamingRule', 'preserve', 'TreatAsEmpty', {'NA', 'N/A', ''});
            s_info = s_info(strcmp(s_info.Modality, 'PET'), :);
        else
            disp('Calculating SUVr requires the subject_info weight parameter');
            return;
        end
    end
    
    % Open the output file.
    fid = fopen(output_csv, 'w');
    % Get the NIfTI file list.
    nii_files = dir(fullfile(pet_dir, [startswith, '*.nii*']));
    

    header = get_header(roi_info, Vox_mm3, SUVr);
    fprintf(fid, '%s\n', strjoin(header, ';'));

    mask = niftiread(mask_path);
    roi_V = spm_vol(mask_path);
    data = spm_read_vols(roi_V); % Read data before changing the name.
    roi_affine  = roi_V.mat;
    unique_values = unique(mask(mask ~= 0));  % Get unique mask values, excluding 0.
    indices = find(data == 1);
    [rows, cols, slices] = ind2sub(size(data), indices);
    % Use the affine matrix if indices need to be converted to physical-space coordinates.
    % Example: get physical coordinates (x, y, z).
    inverse_affine = inv(roi_affine);
    physical_coords = roi_affine * [rows'; cols'; slices'; ones(1, numel(rows))];
    % physical_coords(1, :) is x, physical_coords(2, :) is y, and physical_coords(3, :) is z.
    voxel_coords = inverse_affine * [physical_coords];
    % intensity = nii(sub2ind(size(nii), rows, cols, slices));
    for i = 1:length(nii_files)

        ce_suv = 1;
        weight = -1;

        nii_filename = nii_files(i).name;
        nii_path = fullfile(pet_dir, nii_filename);
        nii = niftiread(nii_path);  % Read the NIfTI file.

        indices = find(mask == ci);

        
        % Calculate SUVr.
        if SUVr
            [~, sid] = fileparts(nii_filename);
            sid = sid(length(prefix)+1:end);  % Get subject ID.
            weight_idx = strcmp(s_info.(ID), sid);
            if any(weight_idx)
                weight = s_info.Weight(weight_idx) * 1000;  % Weight in grams.
            else
                disp([sid, ' has no weight information. SUVr may be 0']);
            end
            
            sum_ce = 0;
            for ci = cerebellar_ID
                indices = find(mask == ci);
                ce_vox_in = nii(indices);  % Get voxel values in cerebellar regions.
                sum_ce = sum_ce + sum(ce_vox_in);
            end
            ce_suv = sum_ce / (Injected_dose / weight);
        end
        
        % Calculate the mean value and other features for each ROI.
        average_values = {};
        for value = fieldnames(roi_info)'
            value = value{1};
            avg_intensity = 0;
            vox_mm3 = 0;
            roi_suvr = 0;
            
            if ~ismember(value, unique_values) || (ismember(value, cerebellar_ID) && ~cerebellar)
                continue;
            else
                indices = find(mask == value);
                intensity = nii(indices);
                avg_intensity = mean(intensity);
                sum_intensity = sum(intensity);
                
                vox_mm3 = numel(indices);
                roi_suv = sum_intensity / (Injected_dose / weight);
                roi_suvr = roi_suv / ce_suv;
            end
            
            average_values = [average_values, {num2str(avg_intensity)}];
            if Vox_mm3
                average_values = [average_values, {num2str(vox_mm3)}];
            end
            if SUVr
                average_values = [average_values, {num2str(roi_suvr)}];
            end
        end
        
        % Write to CSV.
        fprintf(fid, '%s,%s\n', nii_filename, strjoin(average_values, ','));
    end
    
    % Close the file.
    fclose(fid);
    disp(['Results written to CSV file: ', output_csv]);
end

function roi_info = load_roi_info(roi_csv, header, delimiter, n)
    if nargin < 4 n=200; end
    if nargin < 3 delimiter = ';';  end
    if nargin < 2 header = true; end
    if nargin < 1 roi_csv=''; end

    roi_info = {};

    if ~isempty(roi_csv)
        fid = fopen(roi_csv, 'r');
        % Skip the header if present.
        if header
            fgetl(fid);
        end
        
        tline = fgetl(fid);
        while ischar(tline)
            row = strsplit(tline, delimiter);
            roi_value = str2double(row{1});
            if length(row) < 3 
                row{3} = row{2};
            end
            if length(row) < 4 
                row{4} = row{1};
            end
    
            roi_info{end+1} = {roi_value, row{2}, row{3}, row{4}};
            tline = fgetl(fid);
        end
        fclose(fid);
    else
        for i=1:n
            roi_info{end+1} = {i, ['roi_',i], ['roi_',i], i};
        end
    end 
end

function header = get_header(roi_info, Vox_mm3, SUVr)
    header = {'Subject'};
    for i = 1:numel(roi_info)
        roi_ = roi_info{i};
        roi_name = roi_{2};
        header = [header, roi_name];
        if Vox_mm3
            header = [header, [roi_name, '_Vox_mm3']];
        end
        if SUVr
            header = [header, [roi_name, '_SUVr']];
        end
    end
end
